src/many-modules/src/_5_data/types.rs

```diff
@@ -0,0 +1,83 @@
+use many_types::AttributeRelatedIndex;
+use minicbor::{Encode, Decode};
+use num_bigint::BigInt;
+
+pub type DataIndex = AttributeRelatedIndex;
+
+#[derive(Clone, Debug, Decode, Encode, Eq, PartialEq)]
+pub enum DataType {
+    #[n(0)]
+    Counter,
+    #[n(1)]
+    Gauge,
+}
+
+#[derive(Clone, Decode, Encode, PartialEq, Eq, Debug)]
+pub enum DataValue {
+    #[n(0)]
+    Counter(#[n(0)] DataValueTypeCounter),
+    #[n(1)]
+    Gauge(#[n(0)] DataValueTypeGauge),
+}
+
+pub type DataValueTypeCounter = u64;
+
+#[derive(Clone, Decode, Encode, Debug)]
+pub enum DataValueTypeGauge {
+    #[n(0)]
+    Int(#[n(0)] i64),
+    #[n(1)]
+    Float(#[n(0)] f64),
+    #[n(2)]
+    BigInt(#[cbor(n(0), decode_with="decode_bigint", encode_with = "encode_bigint")] BigInt)
+}
+
+impl PartialEq<DataValueTypeGauge> for DataValueTypeGauge {
+    fn eq(&self, other: &DataValueTypeGauge) -> bool {
+        match (self, other) {
+            (DataValueTypeGauge::BigInt(a),
+             DataValueTypeGauge::Int(b)) => &BigInt::from(*b) == a,
+            (DataValueTypeGauge::BigInt(a),
+             DataValueTypeGauge::Float(b)) => b.fract() != 0.0 && &BigInt::from(*b as i64) == a,
```

Comment:
I'll comment on it only once here.

As a good practice, *never* compare two floats for (non-)equality. Compare them using an epsilon, i.e., `fabs(a-b) < eps`.

Fun read: [What Every Computer Scientist Should Know About Floating-Point Arithmetic](https://dl.acm.org/doi/10.1145/103162.103163)

Reply:
True, I'm surprised the Rust standard library does this too.

Reply:
@fmorency but in this case, I'm actually implementing `PartialEq`, which is supposed to be false when comparing non-identical floats. Only if I'm actually comparing in the actual code I should use an epsilon (leaving this job to the user).

Reply:
@Denommus :)

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=95e4215fc1e6b76f619f4f697b0bd5f6